packages/dd-trace/src/plugins/util/inferred_proxy.js

```diff
@@ -0,0 +1,120 @@
+const log = require('../../log')
+const tags = require('../../../../../ext/tags')
+
+const RESOURCE_NAME = tags.RESOURCE_NAME
+const HTTP_ROUTE = tags.HTTP_ROUTE
+const SPAN_KIND = tags.SPAN_KIND
+const HTTP_URL = tags.HTTP_URL
+const HTTP_METHOD = tags.HTTP_METHOD
+
+const PROXY_HEADER_SYSTEM = 'x-dd-proxy'
+const PROXY_HEADER_START_TIME_MS = 'x-dd-proxy-request-time-ms'
+const PROXY_HEADER_PATH = 'x-dd-proxy-path'
+const PROXY_HEADER_HTTPMETHOD = 'x-dd-proxy-httpmethod'
+const PROXY_HEADER_DOMAIN = 'x-dd-proxy-domain-name'
+const PROXY_HEADER_STAGE = 'x-dd-proxy-stage'
+
+const supportedProxies = {
+  'aws-apigateway': {
+    spanName: 'aws.apigateway',
+    component: 'aws-apigateway'
+  }
+}
+
+function createInferredProxySpan (headers, childOf, tracer, context) {
+  if (!headers) {
+    return null
+  }
+
+  if (!tracer._config?.managedServicesEnabled) {
+    return null
+  }
+
+  const proxyContext = extractInferredProxyContext(headers)
+
+  if (!proxyContext) {
+    return null
+  }
+
+  const proxySpanInfo = supportedProxies[proxyContext.proxySystemName]
+
+  log.debug(`Successfully extracted inferred span info ${proxyContext} for proxy: ${proxyContext.proxySystemName}`)
+
+  const span = tracer.startSpan(
+    proxySpanInfo.spanName,
+    {
+      childOf,
+      type: 'web',
+      startTime: proxyContext.requestTime,
+      tags: {
+        service: proxyContext.domainName || this.serviceName(),
+        component: proxySpanInfo.component,
+        [SPAN_KIND]: 'internal',
+        [HTTP_METHOD]: proxyContext.method,
+        [HTTP_URL]: proxyContext.domainName + proxyContext.path,
+        [HTTP_ROUTE]: proxyContext.path,
+        stage: proxyContext.stage,
+        type: 'web'
+      }
+    }
+  )
+
+  tracer.scope().activate(span)
+  context.inferredProxySpan = span
```

Comment:
What class is a `context`? Should we be adding `inferredProxySpan` as a property on that class? E.g. for V8 performance? /cc @bengl

Reply:
The context is just an object that we create and associate with the web request. It's not a `SpanContext` object. When we receive a request we populate an object with the request, span, middlewares used, request path, and some other info, which we can then grab later as long as we have the request.